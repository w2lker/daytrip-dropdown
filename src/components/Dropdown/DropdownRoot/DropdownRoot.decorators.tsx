import withStyles from "react-jss";

import DropdownRoot from "./DropdownRoot";
import DropdownLang from "./DropdownRoot.lang";

import withControlledSwitcher from "../../Decorators/withControlledSwitcher";
import withLangContext from "../../Decorators/withLangContext";

import dropdownRootStyles from "./DropdownRoot.styles";
import lang from "../../../const/lang";

const styledComponent = withStyles(dropdownRootStyles)(DropdownRoot);
const withControlledSwitcherDecorated = withControlledSwitcher(styledComponent);
const withLangContextDecorated = withLangContext(withControlledSwitcherDecorated, DropdownLang, lang.dropdown);

export default withLangContextDecorated;
