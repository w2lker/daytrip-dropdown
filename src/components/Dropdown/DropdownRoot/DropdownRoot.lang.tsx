import {createContext} from "react";

const DropdownLang = createContext({});

export default DropdownLang;
