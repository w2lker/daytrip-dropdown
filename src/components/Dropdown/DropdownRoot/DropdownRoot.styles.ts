const dropdownRootStyles = {
    root: {

    },
};

export default dropdownRootStyles;
