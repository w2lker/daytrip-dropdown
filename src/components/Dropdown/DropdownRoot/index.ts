import Dropdown from "./DropdownRoot.decorators";

export default Dropdown;
