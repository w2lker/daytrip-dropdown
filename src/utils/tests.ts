import { ReactWrapper } from "enzyme";

export function findByDisplayNameRegex(component: ReactWrapper, regExp: RegExp) {
  let child = component.childAt(0);
  while(child.length && !regExp.test(child.name())) {
    child = child.childAt(0);
  }
  return child.length !== 0;
}
