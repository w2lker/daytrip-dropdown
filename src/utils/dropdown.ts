import {IDropdownOptionsElement} from "../components/Dropdown/DropdownRoot/DropdownRoot";

export function getDropdownOptionKey(item: IDropdownOptionsElement): string {
  if (!item) return '';
  return Object.keys(item)[0];
}

export function getDropdownOptionValue(item: IDropdownOptionsElement): string {
  if (!item) return '';
  return Object.values(item)[0];
}

interface IDropdownElementEntries {
  key: string;
  value: string;
}

export function getDropdownOptionEntries(item: IDropdownOptionsElement): IDropdownElementEntries {
  if (!item) return {key: '', value: ''};
  const [key, value] = Object.entries(item)[0];
  return {key, value};
}
