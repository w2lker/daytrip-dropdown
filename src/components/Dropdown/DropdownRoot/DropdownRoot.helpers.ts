import {IDropdownOptionsArray} from "./DropdownRoot";

import {sendWarning} from "../../../utils/warn";
import {getEntries} from "../../../utils/dropdown";

export function modifyDropdownOptionsDuplicates(opts: IDropdownOptionsArray): IDropdownOptionsArray {
    let collector: { [key: string]: string } = {};

    if (!opts) return [];

    collector = opts.reduce( (collection, item) => {
        const keys = Object.keys(item);
        if (keys.length > 1) {
            sendWarning(`Dropdown options should contain only one pair key-value per array item. For element ${item} it skipped some elements`);
        }

        const {key, value} = getEntries(item);
        if (!collection[key]) collection[key] = value;

        return collection;
    }, collector);

    const collectedKeys = Object.keys(collector);
    if (collectedKeys.length === opts.length) {
        return opts;
    }

    sendWarning('Dropdown options should contain unique elements. Some of options are skipped to provide component appropriate functionality');

    return collectedKeys
        .map( (key) => ({
            [key]: collector[key],
        })
    );
}

export function modifyDropdownOptionsMultiline(opts: IDropdownOptionsArray): IDropdownOptionsArray {
    return opts.map( (opt) => {
        const {key, value} = getEntries(opt);
        return {[key]: value.replace('\n', '')}
    })
}
