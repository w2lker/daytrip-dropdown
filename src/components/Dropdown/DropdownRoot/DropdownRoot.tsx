import React, {createContext, useEffect, useState} from 'react';

import DropdownHead from "../DropdownHead";
import DropdownBody from "../DropdownBody";

import {modifyDropdownOptionsDuplicates, modifyDropdownOptionsMultiline} from './DropdownRoot.helpers';
import {getKey} from "../../../utils/dropdown";
import lang from "../../../const/lang";

export interface IDropdownOptionsElement {
    [value: string]: string;
}

export type IDropdownOptionsArray = IDropdownOptionsElement[];

interface IDropdownProps {
    options: IDropdownOptionsArray;
    selected?: string;
    rows?: number;
    label?: string;
    placeholder?: string;
    onSelect?: (selected: string) => void;
    // TODO: define class definition in JSS package
    classes?: any;
}

export const DropdownLang = createContext({});

const DropdownRoot: React.FC<IDropdownProps> = (props) => {
    const { options, rows, label, placeholder, selected, classes, onSelect } = props;
    let modifiedOptions: IDropdownOptionsArray;
    // Filter options array to display only unique key value to prevent issue with onSelect callback validation
    useEffect(() => {
        const withoutDuplicates = modifyDropdownOptionsDuplicates(options);
        modifiedOptions = modifyDropdownOptionsMultiline(withoutDuplicates);
    }, [options]);

    const [isOpened, setIsOpened] = useState(false);

    const selectedOption = selected && options && options.length ?
        options.find( (opt) => getKey(opt) === selected)
    : null;

    return (
      <DropdownLang.Provider value={lang.dropdown}>
        <div className={classes.root}>
            <DropdownHead
                opened={isOpened}
                label={label}
                placeholder={placeholder}
                onClick={() => setIsOpened(!isOpened)}
                selectedItem={selectedOption}
            />
            {isOpened && (
                <DropdownBody
                  selected={selected}
                  options={options}
                  rows={rows}
                  onSelect={onSelect}
                  onClose={() => setIsOpened(false)}
                />
            )}
        </div>
      </DropdownLang.Provider>
    );
};

DropdownRoot.defaultProps = {
    options: [],
    rows: 5,
    classes: {
        root: '',
    }
};

export default DropdownRoot;
