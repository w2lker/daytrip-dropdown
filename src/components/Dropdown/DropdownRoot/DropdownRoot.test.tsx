import React from 'react';

import { mount } from "enzyme";

import DropdownRoot, {IDropdownOptionsArray, IDropdownProps} from "./DropdownRoot";
import dropdownRootStyles from "./DropdownRoot.styles";
import DropdownRootDecorated from './DropdownRoot.decorators';

import {getEntries} from "../../../utils/dropdown";
import {findByDisplayNameRegex} from "../../../utils/tests";
import * as DropdownRootHelpers from "./DropdownRoot.helpers";
const { modifyDropdownOptionsDuplicates, modifyDropdownOptionsMultiline } = DropdownRootHelpers;

describe('DropdownRoot helpers', () => {
  it('modifyDropdownOptionsDuplicates works', () => {
      const testArray: IDropdownOptionsArray = [
        {'key1': 'value1'},
        {'key1': 'value1'},
        {'key1': 'value1'},
        {'key2': 'value2'},
      ];
      const outputArray = modifyDropdownOptionsDuplicates(testArray);
      expect(outputArray.length).toBe(2);
  });
  it('modifyDropdownOptionsMultiline works', () => {
    const expectedArray: IDropdownOptionsArray = [
      {'key1': 'value1 value1 value1 value1'},
      {'key2': 'value1 value1 value1 value1'},
    ];
    const inputArray = expectedArray.map( item => {
      const {key, value} = getEntries(item);
      return {[key]: value.replace(' ', ' \n')}
    });
    const outputArray = modifyDropdownOptionsMultiline(inputArray);
    outputArray.forEach( (item, index) => {
      expect(item).toEqual(expectedArray[index]);
    });
  });
});

const setup = () => {

  const classes = {
    root: 'dropdownRoot-root',
  };
  const classesKeys = Object.keys(classes);

  const sampleProps: IDropdownProps = {
    options: [
      {'opt1': 'value1'},
      {'opt2': 'value2'},
      {'opt3': 'value3'},
      {'opt4': 'value4'},
      {'opt5': 'value5'},
      {'opt6': 'value6'},
      {'opt7': 'value7'},
    ],
    onSelect: jest.fn(),
    classes,
    label: 'some-label',
    placeholder: 'some-placeholder',
    caseInsensitiveSearch: true,
  };

  return {sampleProps, classes, classesKeys};
};

describe('DropdownRoot component', () => {
  it('match snapshot', () => {
      const { sampleProps } = setup();
      const component = mount(<DropdownRoot {...sampleProps} />);
      expect(component.debug()).toMatchSnapshot();
  });
  it('fires helpers on receiving options array', () => {
    const duplicatesSpy = jest.spyOn(DropdownRootHelpers, 'modifyDropdownOptionsDuplicates')
      .mockImplementation( (val) => val );
    const multilineSpy = jest.spyOn(DropdownRootHelpers, 'modifyDropdownOptionsMultiline')
      .mockImplementation( (val) => val );
    const { sampleProps } = setup();
    const component = mount(<DropdownRoot {...sampleProps} />);
    // Initial render
    expect(DropdownRootHelpers.modifyDropdownOptionsDuplicates).toBeCalledTimes(1);
    expect(DropdownRootHelpers.modifyDropdownOptionsMultiline).toBeCalledTimes(1);
    const newOptions = [...sampleProps.options, {'opt3': 'value3'}];
    const newProps = {
      ...sampleProps,
      options: newOptions,
    };
    // @ts-ignore
    component.setProps(newProps);
    // Component Did receive props
    expect(DropdownRootHelpers.modifyDropdownOptionsDuplicates).toBeCalledTimes(2);
    expect(DropdownRootHelpers.modifyDropdownOptionsMultiline).toBeCalledTimes(2);
    duplicatesSpy.mockRestore();
    multilineSpy.mockRestore();
  });
  it('renders head and body', () => {
    const { sampleProps } = setup();
    const component = mount(<DropdownRoot {...sampleProps} />);
    const head = component.find('DropdownHead');
    const body = component.find('DropdownBody');
    expect(head.length).toBe(1);
    expect(body.length).toBe(1);
  })
});

describe('DropdownRoot styles', () => {
  it('contains principle classes', () => {
    const { classesKeys } = setup();
    expect(dropdownRootStyles).toBeDefined();
    classesKeys.forEach((keyValue) => {
      // @ts-ignore
      expect(dropdownRootStyles[keyValue]).toBeDefined();
    });
  });
});

describe('Dropdown decorators', () => {
  it('provides lang context', () => {
    const { sampleProps } = setup();
    const component = mount(<DropdownRootDecorated {...sampleProps} />);
    const langContextReg = new RegExp('^WithLangContext.*');
    expect(findByDisplayNameRegex(component, langContextReg)).toBeTruthy();
  });

  it('provides withControlledSwitch decorator', () => {
    const { sampleProps } = setup();
    const component = mount(<DropdownRootDecorated {...sampleProps} />);
    const langContextReg = new RegExp('^WithControlledSwitcher.*');
    expect(findByDisplayNameRegex(component, langContextReg)).toBeTruthy();
  });

  it('provides styled classes from decorators', () => {
    const { sampleProps, classesKeys } = setup();
    const component = mount(<DropdownRootDecorated {...sampleProps} />);
    // @ts-ignore
    const assignedClasses = component.find('DropdownRoot').props().classes;
    classesKeys.forEach(keyValue => {
      expect(assignedClasses[keyValue]).toBeDefined();
    });
  });
});
