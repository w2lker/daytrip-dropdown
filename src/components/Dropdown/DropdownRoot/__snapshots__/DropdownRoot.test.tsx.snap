// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownRoot component match snapshot 1`] = `
"<DropdownRoot options={{...}} onSelect={[Function: mockConstructor]} classes={{...}} label=\\"some-label\\" placeholder=\\"some-placeholder\\" rows={5}>
  <div className=\\"dropdownRoot-root\\">
    <JssContextSubscriber opened={false} label=\\"some-label\\" placeholder=\\"some-placeholder\\" onClick={[Function: onClick]} selectedItem={{...}}>
      <WithStyles(DropdownHead) innerRef={{...}} opened={false} label=\\"some-label\\" placeholder=\\"some-placeholder\\" onClick={[Function: onClick]} selectedItem={{...}} jssContext={{...}} theme={{...}}>
        <DropdownHead opened={false} label=\\"some-label\\" placeholder=\\"some-placeholder\\" onClick={[Function: onClick]} selectedItem={{...}} classes={{...}}>
          <div className=\\"DropdownHead-wrapper-0-2-1\\" onClick={[Function: onClick]}>
            <div className=\\"DropdownHead-label-0-2-2\\">
              some-label
            </div>
            <div className=\\"DropdownHead-content-0-2-3\\">
              some-placeholder
            </div>
            <div className=\\"DropdownHead-caret-0-2-4\\">
              <svg width=\\"12px\\" height=\\"8px\\" viewBox=\\"0 0 12 8\\">
                <polygon fill=\\"#617E8D\\" points=\\"1.41 0 6 4.59 10.59 0 12 1.42 6 7.42 0 1.42\\" />
              </svg>
            </div>
          </div>
        </DropdownHead>
      </WithStyles(DropdownHead)>
    </JssContextSubscriber>
    <JssContextSubscriber opened={false} selected={[undefined]} options={{...}} rows={5} onSelect={[Function: mockConstructor]} onClose={[Function: onClose]}>
      <WithStyles(DropdownBody) innerRef={{...}} opened={false} selected={[undefined]} options={{...}} rows={5} onSelect={[Function: mockConstructor]} onClose={[Function: onClose]} jssContext={{...}} theme={{...}}>
        <DropdownBody opened={false} selected={[undefined]} options={{...}} rows={5} onSelect={[Function: mockConstructor]} onClose={[Function: onClose]} classes={{...}} />
      </WithStyles(DropdownBody)>
    </JssContextSubscriber>
  </div>
</DropdownRoot>"
`;
